grace-core/src/Grace/Interpret.hs

```diff
@@ -72,31 +69,54 @@ interpretWith
     -> Input
     -> m (Type Location, Value)
 interpretWith bindings maybeAnnotation input = do
-    code <- case input of
-        Path file -> liftIO (Text.IO.readFile file)
-        Code text -> return text
+    let directory = case input of
+            Path file -> FilePath.takeDirectory file
+            Code _ _  -> "."
 
-    let name = case input of
-            Path file -> file
-            Code _    -> "(input)"
+    eitherExpression <- Except.runExceptT (Parser.parseInput input)
 
-    expression <- case Parser.parse name code of
-        Left message -> do
-            Except.throwError (ParseError message)
+    expression <- case eitherExpression of
+        Left e -> throwError (ParseError e)
+        Right expression -> return expression
 
-        Right expression -> do
-            let locate offset = Location{..}
+    interpretExprWith bindings maybeAnnotation directory expression
 
-            return (first locate expression)
+{- | Like `interpretWith`, but takes a pre-parsed expression instead of some
+     source code
+-}
+interpretExprWith
+    :: (MonadError InterpretError m, MonadIO m)
+    => [(Text, Type Location, Value)]
+    -- ^ @(name, type, value)@ for each custom binding
+    -> Maybe (Type Location)
+    -- ^ Optional expected type for the input
+    -> FilePath
+    -- ^ The base directory used when resolving filepath imports of relative paths
+    -> Syntax Location Import
+    -> m (Type Location, Value)
+interpretExprWith bindings maybeAnnotation directory expression = do
+    resolvedExpression <- for (annotate expression)
+        \(maybeAnnotation', input) -> case input of
+            Import.File file -> interpretWith bindings maybeAnnotation' (Path path)
+              where
+                path = FilePath.normalise (directory </> file)
+
+            Import.URI uri -> do
+                eitherResult <- liftIO do
+                    tryAny (Import.resolverToCallback Resolver.defaultResolver uri)
+
+                (importExpression, directory') <- case eitherResult of
+                    Left e -> throwError (ImportError uri (displayException e))
```

Comment:
I think you will want to store the exception and not the rendered form of the exception, like this:

```suggestion
                    Left e -> throwError (ImportError uri e)
```

The main reason I suggest this is to preserve as much structure as possible from the original exception.  Instead, render the underlying exception at the last moment (i.e. have the `displayException` method of `ImportError` call `displayException` for the wrapped exception).

The second reason that I suggest this is that you probably want the `Show` instance for `ImportError` to `Show` (and not `display…`) the underlying exception

Reply:
I totally agree with you; The reason I render the message already there is the `Eq` instance of `ImportError`. Sadly we have to choose between one of the two. What is your preference?

Reply:
I think this is still doable if you use `ExistentialQuantification`, like this:

```haskell
data InterpretError
    = forall e . (Exception e, Eq e) => ImportError URI e
    | …
```

Reply:
The problem here is that we use `tryAny` to catch the exception, hence `e :: SomeException` which does not have an `Eq` instance. We could however change `InterpretError` to
```haskell
data InterpretError
    = ImportError URI SomeException
```
and define the `Eq` instance like
```haskell
instance Eq InterpretError where
    ImportError uri1 e1 == InterpretError uri2 e2 = uri1 == uri2 && show e1 == show e2
```

It is not a pretty "solution" but since the `Eq` instance is only intended to be used in the test suite it _could_ work.

As an alternative we can add another constructor to `Grace.Import.ImportError`:
```haskell
data ImportError
    = forall e. (Eq e, Exception e) => SomeImportError e
    | UnsupportedURI
```
Then we leave it to the resolvers to wrap their errors (and the one of the libraries they are using) and catch only this error in `interpretExprWith`.